octue/cloud/deployment/google/deployer.py

```diff
@@ -0,0 +1,262 @@
+        self.name = octue_configuration["name"]
+        self.repository_name = octue_configuration["repository_name"]
+        self.repository_owner = octue_configuration["repository_owner"]
+        self.project_name = octue_configuration["project_name"]
+        self.region = octue_configuration["region"]
+
+        # Generated attributes.
+        self.service_id = f"{OCTUE_NAMESPACE}.{service_id or uuid.uuid4()}"
+        self.build_trigger_description = f"Build {self.name} service and deploy it to Cloud Run."
+
+        self._default_image_uri = (
+            f"{DOCKER_REGISTRY_URL}/{self.project_name}/{self.repository_owner}/{self.repository_name}/"
+            f"{self.name}:{self._get_short_head_commit_hash()}"
+        )
+
+        # Optional configuration file entries.
+        self.dockerfile_path = octue_configuration.get("dockerfile_path")
+        self.minimum_instances = octue_configuration.get("minimum_instances", 0)
+        self.maximum_instances = octue_configuration.get("maximum_instances", 10)
+        self.concurrency = octue_configuration.get("concurrency", 80)
+        self.image_uri = octue_configuration.get("image_uri", self._default_image_uri)
+        self.branch_pattern = octue_configuration.get("branch_pattern", "^main$")
+        self.memory = octue_configuration.get("memory", "128Mi")
+        self.cpus = octue_configuration.get("cpus", 1)
+        self.environment_variables = octue_configuration.get("environment_variables", [])
+
+    def deploy(self, no_cache=False, update=False):
+        total_number_of_stages = 4
+
+        with ProgressMessage("Generating Google Cloud Build configuration", 1, total_number_of_stages):
+            self._generate_cloud_build_configuration(with_cache=not no_cache)
+
+        with tempfile.NamedTemporaryFile() as temporary_file:
+            with open(temporary_file.name, "w") as f:
+                yaml.dump(self.cloud_build_configuration, f)
+
+            with ProgressMessage("Creating build trigger", 2, total_number_of_stages) as progress_message:
+                try:
+                    self._create_build_trigger(cloud_build_configuration_path=temporary_file.name)
+                except subprocess.SubprocessError as e:
+                    self._raise_or_ignore_already_exists_error(e, update, progress_message)
+
+            with ProgressMessage("Building and deploying service", 3, total_number_of_stages):
+                self._build_and_deploy_service(cloud_build_configuration_path=temporary_file.name)
+
+        with ProgressMessage("Creating Eventarc Pub/Sub run trigger", 4, total_number_of_stages) as progress_message:
+            try:
+                self._create_eventarc_run_trigger()
+            except subprocess.SubprocessError as e:
+                self._raise_or_ignore_already_exists_error(e, update, progress_message)
+
+        print(f"[SUCCESS] Service deployed - it can be questioned via Pub/Sub at {self.service_id!r}.")
```

Comment:
Should this be a log rather than a print?

Reply:
I could make all of them log messages but wasn't sure if it's better as prints for CLI output? It looks like this currently:

```
octue-app deploy --service-id=009ea106-dc37-4521-a8cc-3e0836064334
[1/4] Generating Google Cloud Build configuration...done.
[2/4] Creating build trigger...done.
[3/4] Building and deploying service...done.
[4/4] Creating and attaching Eventarc Pub/Sub run trigger...done.

```

Reply:
Oh, yes! Sorry yes def better for CLI output in that case